src/factory.rs

```diff
@@ -1,37 +1,82 @@
 use crate::*;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+/// Attempts to load and cache the factory interface for the given WinRT class.
+pub struct FactoryCache<C: RuntimeName, I: ComInterface> {
+    pub shared: AtomicPtr<std::ffi::c_void>,
+    pub _c: PhantomData<C>,
+    pub _i: PhantomData<I>,
+}
+
+impl<C: RuntimeName, I: ComInterface + Default> FactoryCache<C, I> {
+    pub fn call<R, F: FnOnce(&I) -> Result<R>>(&mut self, callback: F) -> Result<R> {
```

Comment:
I'm don't think I fully understand why a callback is necessary instead of just returning `&I`.

Reply:
Because the lifetime of `I` depends on whether the factory may be cached and that is only known at run time. If the factory is agile we cache it and a static reference is available. If it's not agile, we cannot cache it and merely use it before releasing.